### PR TITLE
ENH: Use 'Y' as an alias for end of year

### LINKED ISSUE #9313: ENH date_range yearly frequency flag 'A' is confusing, 'Y' should be accepted
to get a yearly frequency from date range the flag is actually `freq="A"` (for "Annual" i guess) rather than `freq="Y"` for yearly.

It would seem more consistent if it accepted "Y" as well/instead of "A", following the pattern:

hour -> hourly -> H
day -> daily -> D
week -> weekly -> W
year -> yearly -> Y

### PR DESCRIPTION
**Pull request by gfyoung**
Redo of #16958.
Closes #9313.

cc @imih 

Additional components:

- [x] `whatsnew` mention
- [x] doc in `timeseries.rst`
- [x] tests for 'YS' and `date_range`

### CODE REVIEW

**Comment by jreback** [comment]
i would have simply taken the original commit and added things in another commit; no need to tense the original away

**Comment by gfyoung** [comment]
> i would have simply taken the original commit and added things in another commit; no need to rebase the original away

That's fair.  Done.

File: doc/source/timeseries.rst [hunk 1]
```diff
@@ -1092,7 +1092,7 @@ frequencies. We will refer to these aliases as *offset aliases*
     "BQ", "business quarter endfrequency"
     "QS", "quarter start frequency"
     "BQS", "business quarter start frequency"
-    "A", "year end frequency"
+    "A, Y", "year end frequency"
     "BA", "business year end frequency"
     "AS", "year start frequency"
```
Comment (jreback) [inline doc/source/timeseries.rst]: YS is also accepted here

File: doc/source/whatsnew/v0.21.0.txt [hunk 2]
```diff
@@ -40,6 +40,7 @@ Other Enhancements
 - :func:`DataFrame.clip()` and :func:`Series.clip()` have gained an ``inplace`` argument. (:issue:`15388`)
 - :func:`crosstab` has gained a ``margins_name`` parameter to define the name of the row / column that will contain the totals when ``margins=True``. (:issue:`15972`)
 - :func:`DataFrame.select_dtypes` now accepts scalar values for include/exclude as well as list-like. (:issue:`16855`)
+- :func:`date_range` now accepts 'Y' in addition to 'A' as an alias for end of year (:issue:`9313`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.21.0.txt]: and add mention of YS

File: pandas/tests/tseries/test_frequencies.py [hunk 3]
```diff
@@ -305,6 +310,10 @@ def test_period_str_to_code():
     assert (frequencies._period_str_to_code('A') == 1000)
     assert (frequencies._period_str_to_code('A-DEC') == 1000)
     assert (frequencies._period_str_to_code('A-JAN') == 1001)
```
Comment (jreback) [inline pandas/tests/tseries/test_frequencies.py]: tests for YS as well (like AS)

File: pandas/tseries/frequencies.py [hunk 4]
```diff
@@ -422,6 +422,27 @@ def get_period_alias(offset_str):
     return _offset_to_period_map.get(offset_str, None)
 
 
+_pure_alias = {
```
Comment (jreback) [inline pandas/tseries/frequencies.py]: why is this necessary? why is this not added to the ``_offsset_to_period_map`` and creating new code?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/tseries/test_frequencies.py (hunk 3 above)
Comment (gfyoung) [inline pandas/tests/tseries/test_frequencies.py]: Umm...where do we actually test 'AS' in this file?

Comment (gfyoung) [inline pandas/tests/tseries/test_frequencies.py]: I added tests for "AS" in `test_date_range`.

**Comment by gfyoung** [comment]
@jreback : PTAL

File: pandas/tests/indexes/datetimes/test_date_range.py [hunk 5]
```diff
@@ -33,6 +33,31 @@ def test_date_range_gen_error(self):
...
+
+    @pytest.mark.parametrize("freq", ["A", "Y"])
+    def test_end_year_alias(self, freq):
+        # see gh-9313
+        rng = date_range("1/1/2013", "7/1/2017", freq=freq)
+        exp = pd.DatetimeIndex(["2013-12-31", "2014-12-31",
+                                "2015-12-31", "2016-12-31"], freq=freq)
+        tm.assert_index_equal(rng, exp)
+
+    @pytest.mark.parametrize("freq", ["BA", "BY"])
```
Comment (jreback) [inline pandas/tests/indexes/datetimes/test_date_range.py]: move to the class for Business tests (below)

File: pandas/tests/tseries/test_frequencies.py [hunk 6]
```diff
@@ -305,6 +310,10 @@ def test_period_str_to_code():
...
+    assert (frequencies._period_str_to_code('Y') == 1000)
```
Comment (jreback) [inline pandas/tests/tseries/test_frequencies.py]: I would add tests that validate things with 'A', 'YS', 'AS' as well here, similar to Y

File: pandas/tests/tseries/test_frequencies.py [hunk 7]
```diff
@@ -369,6 +382,13 @@ def test_freq_group(self):
         assert frequencies.get_freq_group('-1A') == 1000
         assert frequencies.get_freq_group('A-JAN') == 1000
         assert frequencies.get_freq_group('A-MAY') == 1000
+
+        assert frequencies.get_freq_group('Y') == 1000
+        assert frequencies.get_freq_group('3Y') == 1000
+        assert frequencies.get_freq_group('-1Y') == 1000
+        assert frequencies.get_freq_group('Y-JAN') == 1000
```
Comment (jreback) [inline pandas/tests/tseries/test_frequencies.py]: also BYS (assume BAS is validated somewhere)

File: pandas/tseries/frequencies.py [hunk 8]
```diff
@@ -709,9 +717,20 @@ def get_standard_freq(freq):
...
+        alias = "Y" + k[1:]
+        year_aliases[alias] = v
+
+_period_code_map.update(**year_aliases)
+del year_aliases
+
 _period_code_map.update({
     "Q": 2000,  # Quarterly - December year end (default quarterly)
     "A": 1000,  # Annual
+    "Y": 1000,
```
Comment (jreback) [inline pandas/tseries/frequencies.py]: I don't think you need this, the point is the alias are mapped to here

File: pandas/tseries/frequencies.py [hunk 9]
```diff
@@ -709,9 +717,20 @@ def get_standard_freq(freq):
     _reverse_period_code_map[_v] = _k
 
 # Additional aliases
```
Comment (jreback) [inline pandas/tseries/frequencies.py]: say that these are yearly aliases (Y, A)

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/indexes/datetimes/test_date_range.py (hunk 5 above)
Comment (gfyoung) [inline pandas/tests/indexes/datetimes/test_date_range.py]: But this is for `date_range`, not `bdate_range` though.

Comment (jreback) [inline pandas/tests/indexes/datetimes/test_date_range.py]: hmm ok.

File: pandas/tests/tseries/test_frequencies.py (hunk 6 above)
Comment (gfyoung) [inline pandas/tests/tseries/test_frequencies.py]: Not sure what you mean here: neither `YS` nor `AS` are considered valid "period strings"

File: pandas/tests/tseries/test_frequencies.py (hunk 7 above)
Comment (gfyoung) [inline pandas/tests/tseries/test_frequencies.py]: Not sure what you mean here: neither `BYS` nor `BAS` are considered valid "frequency groups"

File: pandas/tseries/frequencies.py (hunk 9 above)
Comment (gfyoung) [inline pandas/tseries/frequencies.py]: Done.

File: pandas/tseries/frequencies.py (hunk 8 above)
Comment (gfyoung) [inline pandas/tseries/frequencies.py]: Done.

File: pandas/tests/tseries/test_frequencies.py (hunk 6 above)
Comment (jreback) [inline pandas/tests/tseries/test_frequencies.py]: well, where is AS tested then?

Comment (gfyoung) [inline pandas/tests/tseries/test_frequencies.py]: In `date_range` at the very least.

**Comment by gfyoung** [comment]
@jreback : Everything is green.  PTAL

**Comment by jreback** [comment]
lgtm

**Comment by jreback** [comment]
http://pandas-docs.github.io/pandas-docs-travis/timeseries.html#offset-aliases

needs updates for biz start/end freq

**Comment by gfyoung** [comment]
Oops.  Done.

**Pull request merged**